nixos/modules/virtualisation/qcow.nix

```diff
@@ -0,0 +1,57 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.virtualisation.qcow;
+in
+{
+  options = {
+    virtualisation.qcow = {
+      baseImageSize = mkOption {
+        type = with types; either (enum [ "auto" ]) ints.positive;
+        default = "auto";
+        example = 8192;
+        description = ''
+          The base size of the QCOW2 image in MiB.
+        '';
+      };
+      additionalImageSpace = mkOption {
+        type = types.int;
+        default = 512;
+        example = 1024;
+        description = ''
+          Additional disk space in MiB to be added to the image if <literal>virtualisation.qcow.baseImageSize</literal> is set to <literal>auto</literal>.
+        '';
+      };
+      imageBootPartitionSize = mkOption {
+        type = types.int.positive;
+        default = 256;
+        example = 512;
+        description = ''
+          The size of the boot partition in MiB, only takes effect if <literal>virtualisation.qcow.partitionTableType</literal> is set to <literal>efi</literal> or <literalhybrid</literal>.
+        '';
+      };
+      imagePartitionTableType = mkOption {
+        type = types.enum [ "legacy" "efi" "legacy+gpt" "hybrid" "none" ];
+        default = "legacy";
+        example = "efi";
+        description = ''
+          Type of partition table to use for the image, available options are <literal>legacy</literal>, <literal>efi</literal>, <literal>legacy+gpt</literal>, <literal>hybrid</literal> or <literal>none</literal>. For additional information about each option: <link>https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix#L32</link>.
+        '';
+      };
+    };
+  };
+
+  config = {
+    system.build.qcow = import ../../lib/make-disk-image.nix {
+      inherit lib config pkgs;
+      name = "qcow-image";
+      format = "qcow2";
```

Comment:
This is actually an arbitrary choice and another choice would not fit the `qcow` option names.
The module should be generalized somehow. Can be done in a couple of ways:

a. Rename `virtualisation.qcow` to something like `virtualisation.diskImage` and make the type configurable with a `qcow2` default.
b. Alternatively, you could extract a submodule from this and add it add the options `virtualisation.image.qcow`, `virtualisation.image.raw`, etc. This shifts the complexity to the implementation, disproportionally. It is also arbitrary in a new way: why can we have multiple image formats simultaneously for the same config, but no differing sizes, partition tables, etc.
c. Yet another alternative is to use a submodule in `attrsOf` so the user can create multiple images from the same config simultaneously. While this is almost precisely as powerful as calling `make-disk-image.nix` yourself, it is hardly more convenient.

I'd go with (a) for its simplicity.

Reply:
So basically you can do (a): rename and add an option for the format.